packages/highlight-vdom/src/Highlight.tsx

```diff
@@ -0,0 +1,118 @@
+/** @jsx createElement */
+import {
+  cx,
+  ComponentChildren,
+  ComponentProps,
+  ElementType,
+  Renderer,
+} from '@algolia/ui-components-shared';
+
+type HighlightPartProps = {
+  children: ComponentChildren;
+  classNames: Partial<HighlightClassNames>;
+  highlightedTagName: ElementType;
+  nonHighlightedTagName: ElementType;
+  isHighlighted: boolean;
+};
+
+function createHighlightPartComponent({ createElement }: Renderer) {
+  return function HighlightPart({
+    classNames,
+    children,
+    highlightedTagName,
+    isHighlighted,
+    nonHighlightedTagName,
+  }: HighlightPartProps) {
+    const TagName = isHighlighted ? highlightedTagName : nonHighlightedTagName;
+
+    return (
+      <TagName
+        className={
+          isHighlighted ? classNames.highlighted : classNames.nonHighlighted
+        }
+      >
+        {children}
+      </TagName>
+    );
+  };
+}
+
+type HighlightedPart = {
+  isHighlighted: boolean;
+  value: string;
+};
+
+export type HighlightClassNames = {
+  /**
+   * Class names to apply to the root element
+   */
+  root: string;
+  /**
+   * Class names to apply to the highlighted parts
+   */
+  highlighted: string;
+  /**
+   * Class names to apply to the non-highlighted parts
+   */
+  nonHighlighted: string;
+  /**
+   * Class names to apply to the separator between highlighted parts
+   */
+  separator: string;
+};
+
+export type HighlightProps = ComponentProps<'span'> & {
+  classNames?: Partial<HighlightClassNames>;
+  highlightedTagName?: ElementType;
+  nonHighlightedTagName?: ElementType;
+  separator?: ComponentChildren;
+  parts: HighlightedPart[][];
+};
+
+export function createHighlightComponent({
+  createElement,
+  Fragment,
+}: Renderer) {
+  return function Highlight({
+    parts,
+    highlightedTagName = 'mark',
+    nonHighlightedTagName = 'span',
+    separator = ', ',
+    className,
+    classNames = {},
+    ...props
+  }: HighlightProps) {
+    const HighlightPart = createHighlightPartComponent({
```

Comment:
this should be a level higher, before Highlight is returned so it's created once, instead of every render

Reply:
Yes of course, it has nothing to do here >.<

Reply:
Addressed in [faa0660](https://github.com/algolia/ui-components/pull/2/commits/faa06606767f343e85021d6a1760fe833c55e22d).